[mono] Implement IgnoresAccessChecksToAttribute support

lambdageek opened the pull request:
System.Runtime.CompilerServices.IgnoresAccessChecksToAttribute is like the reverse of
InternalsVisibleTo - it's added to an assembly to indicate that that
assembly can call non-public methods from the specified assembly.

The interesting thing is that the DynamicProxy dynamically generates
the attribute, so it is not explicitly declared in CoreLib.

Fixes https://github.com/dotnet/runtime/issues/47989

lambdageek committed:
Implement IgnoresAccessChecksToAttribute support

System.Runtime.CompilerServices.IgnoresAccessChecksToAttribute is like the reverse of
InternalsVisibleTo - it's added to an assembly to indicate that that
assembly can call non-public methods from the specified assembly.

The interesting thing is that the DynamicProxy dynamically generates
the attribute, so it is not explicitly declared in CoreLib.

Fixes https://github.com/dotnet/runtime/issues/47989

CoffeeFlux commented on src/mono/mono/metadata/class.c:
```diff
@@ -5855,6 +5855,26 @@ is_valid_family_access (MonoClass *access_klass, MonoClass *member_klass, MonoCl
+		MonoAssemblyName *victim = (MonoAssemblyName *)tmp->data;
+		if (!victim->name)
+			continue;
+		if (g_ascii_strcasecmp (accessed->aname.name, victim->name))
```
Purely a nit, but isn't this cleaner by just writing `if (!condition) return`? The continue seems odd.

CoffeeFlux commented on src/mono/mono/metadata/assembly.c:
```diff
@@ -2126,7 +2126,11 @@ mono_assembly_load_friends (MonoAssembly* ass)
 	mono_assemblies_unlock ();
 
-	list = NULL;
+	GSList *visible_list;
```
Another nit, but you can just initialize these to NULL on the same line.

lambdageek commented on src/mono/mono/metadata/class.c:
```diff
@@ -5855,6 +5855,26 @@ is_valid_family_access (MonoClass *access_klass, MonoClass *member_klass, MonoCl
+static gboolean
+ignores_access_checks_to (MonoAssembly *accessing, MonoAssembly *accessed)
+{
+	if (accessing == accessed)
+		return TRUE;
```
Hmm.  I dont' think this is right.  I think this will make all `private` members accessible.

CoffeeFlux replied:
Wouldn't `can_access_internals` be broken too then?

lambdageek replied:
No because `can_access_internals` is only called for `internal` or `protected` or `private protected`.  IgnoresAccessChecks is actually more powerful (c.f. the re-enabled tests) it will let you reach in to get `private` methods too.

CoffeeFlux replied:
I would be somewhat surprised if that did not trigger a test failure, but maybe it's disabled for other reasons. If a test like that does not exist, it also might be worth explicitly adding one.

lambdageek replied:
Working on adding one.  I coudln't find aything like it.  There's not a lot of mentions of MethodAccessException in the codebase!

lambdageek replied:
Added 2 more tests.  Now we check that calling `internal`/`private` methods across assemblies throws a `MethodAccessException` and that calling a `private` method in the same assembly also throws.

lambdageek committed:
Don't blindly allow an assembly to access its own private members

lambdageek committed:
style fixes

lambdageek committed:
Add System.Reflection.Emit tests that throw MethodAccessException

check that calling private/internal methods in another assembly throws.
check that calling a private method in the same assembly throws.

krwq commented:
Thank you @lambdageek!
